### PR TITLE
🔧 Tilt auto-detects local flavors from templates dir

### LINKED ISSUE #935: Auto detect available flavors in Tilt
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Currently, using cluster flavors as Tilt resources requires adding them to tile_config.json as described in https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/master/templates/flavors#from-tilt-config. Furthermore, this requires developers to remember adding new flavors to the config whenever a new template is added. 

Desired solution: By default, flavors are automatically fetched from the templates/flavors directory and added as tilt resources. If and only if `worker-flavors` is defined in tilt_config.json, then use that list as tilt resources (to allow overriding default flavors if you want a subset only).

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind other

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #935 

**Special notes for your reviewer**: Best way to test this is probably to checkout the PR branch and run `make tilt-up`.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Tilt auto-detects local flavors from templates dir
```

### CODE REVIEW

**Review by devigned — APPROVED** [review]
/lgtm

/assign @nader-ziada

**Review by nader-ziada — COMMENTED** [review]
I tried running it but got this error a couple of times

```
Successfully loaded Tiltfile (1m1.336802958s)
Traceback (most recent call last):
  /Users/pivotal/workspace/cluster-api-provider-azure/Tiltfile:339:8: in <toplevel>
  /Users/pivotal/workspace/cluster-api-provider-azure/Tiltfile:232:32: in flavors
  /Users/pivotal/workspace/cluster-api-provider-azure/Tiltfile:287:20: in deploy_worker_templates
  /Users/pivotal/workspace/cluster-api-provider-azure/Tiltfile:314:21: in envsubst
  <builtin>: in local
Error: command ["sh" "-c" "echo \"\x00\x00\x00\x01Bud1\x00\x00\x10\x00\x00\x00\b\x00\x00\x00\x10\x00\x00\x00\x02\t\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\b\x00\x00\x00\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
......
0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\" | ./hack/tools/bin/envsubst"] failed.
error: fork/exec /bin/sh: invalid argument
stdout: ""
stderr: ""
```

tilt up works fine in the same terminal env when switching to a different branch without this change

**Comment by CecileRobertMichon** [comment]
@nader-ziada what version of tilt do you have?

**Comment by CecileRobertMichon** [comment]
Also are you able to run hack/tools/bin/envsubst on its own?

**Comment by devigned** [comment]
```shell
$ tilt version
v0.17.10, built 2020-10-30
$ uname -a
Linux tadashi 5.8.0-28-generic #30-Ubuntu SMP Thu Nov 5 13:24:33 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
Seems to be working for me.

**Comment by nader-ziada** [comment]
> ```shell
> $ tilt version
> v0.17.10, built 2020-10-30
> $ uname -a
> Linux tadashi 5.8.0-28-generic #30-Ubuntu SMP Thu Nov 5 13:24:33 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
> ```
> 
> Seems to be working for me.

my tilt is `v0.17.6, built 2020-09-29` will upgrade and try again

**Comment by nader-ziada** [comment]
I upgraded tilt v0.17.11 and still got the same issues  

> Also are you able to run hack/tools/bin/envsubst on its own?
yes

if this is a problem just on my environment, we can go ahead and merge and I'll figure it out later

**Comment by CecileRobertMichon** [comment]
@nader-ziada we should figure it out, if you're running into it others might too

Are you on MacOS? did you try deleting hack/ envsubst ?

**Comment by CecileRobertMichon** [comment]
Also what docker version are you on?

**Comment by nader-ziada** [comment]
> Also what docker version are you on?

docker 19.03.13

**Comment by nader-ziada** [comment]
turned out the problem was that I had a `.DS_Store` file in the templates folder which was causing the `envsubst` to give an error, once I deleted that, tilt worked

File: Tiltfile [hunk 1]
```diff
@@ -224,23 +224,23 @@ def flavors():
...
         os.environ.update({ssh_pub_key: base64_encode_file(ssh_pub_key_path)})
 
-    for flavor in cfg.get("worker-flavors", []):
-        if flavor not in worker_templates:
-            worker_templates.append(flavor)
-    for flavor in worker_templates:
-        deploy_worker_templates(flavor, substitutions)
+    templatelist = [ item for item in listdir("./templates") ]
+    for template in templatelist:
+        if template not in worker_templates and os.path.basename(template).startswith("cluster-template"):
```
Comment (CecileRobertMichon) [inline Tiltfile:229]: @nader-ziada `and os.path.basename(template).startswith("cluster-template")` should fix the issue you ran into when another file is present

Comment (nader-ziada) [reply Tiltfile:229]: working now, even with the .DS_Store file there :)

**Comment by nader-ziada** [comment]
thanks for figuring out the problem and the quick fix @CecileRobertMichon 

/lgtm
/approve

**Pull request merged**